Providers/sql_server_login/scripts/sql_server_provider.prerequisites.tests.ps1

```diff
@@ -1,127 +1,204 @@
+
+            $process = New-Object System.Diagnostics.Process
+            $process.StartInfo = $processInfo
+            $process.Start() | Out-Null
+            $process.WaitForExit()
+            $errorOutput = $process.StandardError.ReadToEnd()
+
+            !$errorOutput
+            
+        }
+        'ParametersUsed' = @('WindowsAccountCredential')
+    },
+    @{
+        'Name'           = 'the Windows account has permission to update SQL login passwords'
+        'Command'        = {
+            
+            $testCode = {
+                param($Endpoint, $Port)
                 $connection = New-Object System.Data.SqlClient.SqlConnection
-                $connection.ConnectionString = "Server=$ProviderEndpoint\.,$Port;Database=master;Integrated Security=True;"
+                $connection.ConnectionString = ('Server={0},{1};Database=master;Integrated Security=True;' -f $Endpoint, $Port)
                 try {
                     $connection.Open()
                 } catch {
-                    Set-ItResult -Inconclusive
+                    Write-Output 'inconclusive'
                 }
-
                 $command = $connection.CreateCommand()
                 $command.CommandText = "SELECT CASE WHEN IS_SRVROLEMEMBER('sysadmin') = 1 OR IS_SRVROLEMEMBER('securityadmin') = 1 OR IS_ROLEMEMBER('db_owner') = 1 THEN 1 ELSE 0 END"
             
                 $command.ExecuteScalar() | Should -Be 1
                 $connection.Close()
             }
+
+
+            $processInfo = New-Object System.Diagnostics.ProcessStartInfo
+            $processInfo.FileName = "pwsh.exe"
+            $processInfo.Arguments = "-NoProfile -Command & {$($testCode.ToString())} -Endpoint '$Endpoint' -Port $Port"
+            $processInfo.UseShellExecute = $false
+            $processInfo.RedirectStandardOutput = $true
+            $processInfo.LoadUserProfile = $false
+            $processInfo.RedirectStandardError = $true
+            $processInfo.CreateNoWindow = $true
+            $processInfo.Domain = $cred.GetNetworkCredential().Domain
+            $processInfo.UserName = $cred.GetNetworkCredential().UserName
+            $processInfo.Password = $cred.Password
+
+            $process = New-Object System.Diagnostics.Process
+            $process.StartInfo = $processInfo
+            $process.Start() | Out-Null
+            $process.WaitForExit()
+            $errorOutput = $process.StandardError.ReadToEnd()
+
+            !$errorOutput
         }
+        'ParametersUsed' = @('WindowsAccountCredential')
+    }
+)
+
+[array]$passedTests = foreach ($test in $tests.where({ $paramsUsed = $_.ParametersUsed; !$_.ContainsKey('ParametersUsed') -or $PSBoundParameters.Keys.where({ $_ -in $paramsUsed }) })) {
+    $result = & $test.Command
+    if (-not $result) {
+        Write-Error -Message "The test [$($test.Name)] failed."
+    } else {
+        1
     }
-}
\ No newline at end of file
+}
+
+if ($passedTests.Count -eq $tests.Count) {
```

Comment:
I think the result will always be "Some tests failed. Please check the errors above." since you will not run all tests (you will run for WindowsAccountCredential or SqlLoginCredential)
One way to fix it would be to add
`$testsToRun = $tests.where({ $paramsUsed = $_.ParametersUsed; !$_.ContainsKey('ParametersUsed') -or $PSBoundParameters.Keys.where({ $_ -in $paramsUsed }) })`

You could check if passedTests.Count -eq $testsToRun.Count

Am I right ? (not an expert in powershell)

Reply:
I did find a few issues with the way I had the tests structured. I've improved the overall framework and fixed the tests. It _should_ work now.

Reply:
From my understanding, if everything works well, I should get the message 'All tests have passed. You're good to go!' 
But I tried and didn't get the message. No error in the test and the message of success was not there.  If I enter the wrong password, the test of authentication fail and that makes sense. But if everything works, I don't get the success message.

Am I wrong or the $passedTests.Count will never be equal to $tests.Count ?  The $tests.Count will always be 5 and the $passedTests is 3 in my case since I use SQL authentication.

Reply:
Ah, I see that now. I'll add a fix for that.